ONSdigital/sds
# build & push image to dockerhub on pull request for specific branches

### Motivation and Context
Currently the service will create and push images to a Google cloud registry within the project. This will become an issue if the service is destroyed for any reason and recreated.

To safeguard and build robust image creation, work can start to use github workflows to create and push the images to docker hub for defined branch names used in active development, this will create images that can be manually used or in a build trigger for development and or testing of hotfixes and bugs etc. 

### What has changed

1. Added new github workflow file 

- to publish image with latest tag on dockerhub on merge to main  => publish-latest-image-on-merge.yml
- to publish image with version tag on dockerhub on release to main  => publish-image-on-release.yml

2.  Updated cloudbuild-prod & cloudbuild-preprod to push image to ons-sds-dns GAR

3. Added IAM policy to ons-sds-dns
PreProd:
- service-408734099180@serverless-robot-prod.iam.gserviceaccount.com (Artifact Registry Reader)
- ons-sds-cloudbuild-sa@ons-sds-preprod.iam.gserviceaccount.com (Artifact Registry Service Agent, Artifact Registry Writer)
Prod:
- **Missing ons-sds-prod serverless-robot-prod account**
- ons-sds-cloudbuild-sa@ons-sds-prod.iam.gserviceaccount.com (Artifact Registry Service Agent, Artifact Registry Writer)

### How to test?
- Check for pipeline to pass all the test
- Check that the image is pushed to dockerhub onsdigital/sds repo
- Check image with tag in ons-sds-dns GAR

### Links
[Jira issue](https://jira.ons.gov.uk/browse/SDSS-196)
[dockerhub repo](https://hub.docker.com/r/onsdigital/sds/tags)

### Screenshots
Image pushed to DockerHub (Tag for testing: v0.0.42)
![image](https://github.com/ONSdigital/sds/assets/129969878/6860b8e3-2fb9-47d8-aaa5-6dc04a78037e)
Image uploaded on ons-sds-dns GAR with tag :latest and :v0.0.42
![image](https://github.com/ONSdigital/sds/assets/129969878/0e03392e-e1cc-4e8f-8bfc-40b1f872011e)
Image deployed to Preprod from ons-sds-dns GAR
![image](https://github.com/ONSdigital/sds/assets/129969878/356bf675-0103-432b-8f4b-4c02ab83ab57)
